### PR TITLE
[Filtered Mode:] Avoid missed diagnostics for deletion-only hunks

### PR DESCRIPTION
**Pull request by WodansSon**
## Summary

This change keeps filtered mode focused on PR-introduced issues, while fixing the gap where deletion-only hunks could cause us to miss diagnostics that were actually affected by the change.

The main architectural change is that filtering is no longer decided ad hoc inside each analyzer. Analyzers now report diagnostics with explicit metadata about the evidence for that diagnostic, and the runner applies the final filtering decision centrally.

## Problem

The original delete-hunk work fixed one real issue, but the naive versions of that fix had an unacceptable tradeoff:

- if we fell back to "the file changed, so report it", filtered mode could start surfacing unrelated pre-existing issues
- if we stayed with exact line matching only, deletion-only hunks could still miss diagnostics for structural rules where the reported position is not the same thing as the changed evidence

The requirement for this PR is stricter than "catch more issues":
filtered mode must still hide unrelated pre-existing issues and only surface diagnostics that are meaningfully tied to the current diff.

### Centralized filtering

Filtering now happens in the runner instead of being treated as analyzer-local behavior.

- analyzers emit diagnostics plus metadata about the evidence lines and matching mode
- the runner looks up that metadata before printing a diagnostic
- the loader decides whether the metadata matches the current diff

This moves the real filtering logic into one place and makes the behavior auditable and testable.

### Diff model now understands hunks, not just added lines

The loader now parses full hunk metadata instead of treating the diff as only a set of changed line numbers.

That lets us support:
- exact added-line matching for ordinary rules
- same-hunk matching for structural rules
- file-changed and new-file matching when a rule genuinely needs it

### New reporting metadata path

A shared reporting helper was added so analyzers can describe why a diagnostic should survive filtering.

This metadata includes:
- report position
- evidence file
- evidence lines
- match mode

That gives us enough information to distinguish:
- where a diagnostic is reported
- which changed lines actually justify keeping it

### Structural rules no longer rely on fake line-local matching

Most rules still behave like the current linter: they match exact added lines.

Structural rules now have a better option when exact line matching is not the right model. The main example in this patch is AZNR005, which now uses same-hunk matching so deletion-only changes in a registration block can still surface a sorting diagnostic without widening to the whole file.

### Legacy loader gate removed

The old `ShouldReport` line gate has been removed from the analyzer flow. That old API encoded too little information for the behavior we need.

## Why This Approach

This is the smallest design that meets both constraints at the same time:

1. filtered mode should not surface unrelated pre-existing issues
2. deletion-only hunks should still be able to surface diagnostics when the change actually affected the relevant structure

A single boolean gate like "should report this line" is not expressive enough for that. We need to know what evidence justifies the diagnostic, not just where the analyzer happened to report it.

### What stays the same

- filtered mode still suppresses unrelated pre-existing issues
- most rules still only survive when their evidence is on added lines
- `--no-filter` still reports everything

### What improves

- deletion-only hunks no longer require a broad file-level fallback
- structural analyzers can opt into better matching semantics when exact line matching is too narrow
- filtering behavior is now consistent and centralized instead of being spread across analyzers

## Validation

Validation is green.

- `go test ./...` passes
- runner-level tests were added for:
  - same-hunk filtering
  - exact-added filtering
  - new-file metadata filtering
  - cross-file evidence
  - structural evidence where report position differs from evidence lines
  - missing-evidence fallback to the report location
  - suppression of unrelated AZBP005 header diagnostics for deletion-only diffs
- loader tests were updated to verify exact-added and same-hunk behavior directly against the new filtering model

## Risk / Regression Notes

This PR is intended to preserve the current filtered-mode behavior for ordinary rules while improving correctness for delete hunks and structural diagnostics.

I would not claim "zero possible regressions", but the known high-risk paths are now covered by targeted tests, and one validation-discovered edge case around missing evidence metadata was fixed before finalizing the patch.

## Additional Notes

* The README was also updated to document the new filtered-mode semantics more accurately, since "only changed lines are analyzed" is no longer precise enough to describe the behavior.

* Fix AZNR005 filtered-mode misses when unsorted registration literals are assigned to a local variable before being returned. The analyzer now follows function-local identifier bindings back to the underlying composite literal so ordering checks remain semantic rather than dependent on a direct return <literal> shape. Regression tests were updated to cover the variable-backed case while keeping the fixture orthogonal to AZBP010’s declared-and-immediately-returned rule.

* A follow-up review fix moved `AZNR001` onto metadata-backed new-file filtering so schema-order diagnostics stay correctly scoped in filtered mode, and kept `AZBP002`'s `lintignore` handling aligned with the metadata-based model after the `v0.1.9` rule-name change.

* Follow-up cleanup aligned the new regression tests with CI lint expectations by checking test cleanup errors explicitly and fixing `gofumpt` formatting in the added test coverage. This does not change linter behavior, but keeps the validation path green for the `filtered-mode` refactor.

* A small post-review cleanup tightened the new test scaffolding and temporary directory state so the added runner and analyzer regressions satisfy `errcheck` and formatting rules without changing the runtime semantics of the linter.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,83 +1,89 @@
...
-- AZNR001: check for Schema field ordering
-- AZNR002: check for top-level updatable arguments are included in Update func
-- AZNR003: check for `expand*`/`flatten*` functions are defined as receiver methods
-- AZRE001: check for fixed error strings using `fmt.Errorf` instead of `errors.New`
-- AZRN001: check for percentage properties use `_percentage` suffix instead of `_in_percent`
-- AZSD001: check for `MaxItems:1` blocks with single property should be flattened
-- AZSD002: check for `AtLeastOneOf` or `ExactlyOneOf` validation on TypeList fields with all optional nested fields
+# Changelog
+
+## Unreleased
```
Comment (QixiaLu) [inline CHANGELOG.md:3]: This seems to be outdated. v0.1.9 contains version and output json changes already.

**Review by QixiaLu — COMMENTED** [review]
@WodansSon Thanks for putting this PR! These changes look great! I have tested locally. It works like a charm. Just a small comment about the outdated ChangeLog.md; And there are some golangci-lint issues.

File: CHANGELOG.md (hunk 1 above)
Comment (WodansSon) [reply CHANGELOG.md:3]: Fixed.

**Comment by QixiaLu** [comment]
LGTM, Thanks @WodansSon !
